### PR TITLE
Add automated testing

### PR DESCRIPTION
**Pull request by erlingrj**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -61,3 +61,5 @@ dkms.conf
 # VS Code
 .vscode
 
+# Test results
+test/results/
```
Comment (lhstrh) [inline .gitignore]:
```suggestion
test/results/

```

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	bash $(TEST_SCRIPTS)/test_prolog.sh $@
+	lfc -c $^ 
+	sleep 2
+	bash $(TEST_SCRIPTS)/test_epilogue.sh 
+	bash $(TEST_SCRIPTS)/test_parse.sh $@
+
+.PHONY: clean
+clean:
+	rm -r src-gen
+	rm test/results/*
```
Comment (lhstrh) [inline Makefile]:
```suggestion
	rm test/results/*

```

**Comment by erlingrj** [comment]
This setup is not for testing "interactive" LF programs where the user must press a button or observe LED toggles. It is for testing that Actions, Timers etc work. If the program successfully shutdowns it is a success. If there is a lf_print_error_and_exit, or it doesn't shutdown, it is a failure.

File: test/HelloWorld2.lf [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    build: "../scripts/build_nrf_unix.sh",
+    threading: false,
+    build-type: Debug
+}
+main reactor {
+    reaction(startup) {=
+        // lf_print_error_and_exit("test");
+        printf("Hello world\n");
+    =}
+}
```
Comment (lhstrh) [inline test/HelloWorld2.lf]:
```suggestion
}

```

File: test/TestAction.lf [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    logical action a;
+
+    reaction(startup) -> a {=
+        lf_schedule(a, MSEC(10));
+    =}
+
+    reaction(a) {=
+        printf("Action triggered\n");
+    =}
+}
```
Comment (lhstrh) [inline test/TestAction.lf]:
```suggestion
}

```

File: test/TestBlink.lf [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    build: "../scripts/build_nrf_unix.sh",
+    threading: false,
+    timeout: 1 sec
+}
+
+import FlashingLEDs from "../src/BuiltInLED.lf"
+
+main reactor {
+    leds = new FlashingLEDs()
+}
```
Comment (lhstrh) [inline test/TestBlink.lf]:
```suggestion
}

```

File: test/TestTimer.lf [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+    threading: false,
+    timeout: 1 sec
+}
+main reactor {
+    timer t(0, 100 msec)
+
+    reaction(t) {=
+        printf("Timer triggered\n");
+    =}
+}
```
Comment (lhstrh) [inline test/TestTimer.lf]:
```suggestion
}

```

File: test/TestTimer.lf [hunk 7]
```diff
@@ -0,0 +1,12 @@
+target C {
+    build: "../scripts/build_nrf_unix.sh",
+    threading: false,
+    timeout: 1 sec
+}
+main reactor {
+    timer t(0, 100 msec)
+
+    reaction(t) {=
+        printf("Timer triggered\n");
```
Comment (lhstrh) [inline test/TestTimer.lf:11]: Shouldn't there be a test condition in a shutdown reaction, that checks how many times the reaction was invoked?

File: test/scripts/test_epilogue.sh [hunk 8]
```diff
@@ -0,0 +1,4 @@
+# Stop JLink RTT client
+pkill JLinkRTTClient
+# Stop JLink Client
+pkill JLinkExe
```
Comment (lhstrh) [inline test/scripts/test_epilogue.sh]:
```suggestion
pkill JLinkExe

```

File: test/scripts/test_parse.sh [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+fi
+
+if grep --quiet "ERROR" $1; then
+echo "ERROR: Test assertion failed"
+exit -1
+fi
+
+echo "Test: $1 succeeded"
+
+
```
Comment (lhstrh) [inline test/scripts/test_parse.sh]:
```suggestion
```

File: test/scripts/test_prolog.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+# Start JLink server
+echo STARTING JLINK SERVER
+x-terminal-emulator -e "JLinkExe -device nrf52832_xxaa -if swd -speed 4000 -AutoConnect 1"
+echo STARTING JLINK RTT CLIENT
+JLinkRTTClient > $1 &
```
Comment (lhstrh) [inline test/scripts/test_prolog.sh]:
```suggestion
JLinkRTTClient > $1 &

```

File: test/TestAction.lf [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+main reactor {
+
+    logical action a;
+
+    reaction(startup) -> a {=
+        lf_schedule(a, MSEC(10));
+    =}
+
+    reaction(a) {=
+        printf("Action triggered\n");
```
Comment (lhstrh) [inline test/TestAction.lf]: I would set a state variable here and have a shutdown reaction that checks whether the variable was set.

**Review by lhstrh — CHANGES_REQUESTED** [review]
Looks great! Thanks for putting this together so quickly! I left a few comments.

File: test/TestTimer.lf (hunk 7 above)
Comment (erlingrj) [reply test/TestTimer.lf:11]: Done

**Review by lhstrh — COMMENTED** [review]
:rocket:

**Pull request merged**